test_plans/reference_semantics.asciidoc

```diff
@@ -0,0 +1,148 @@
+
+- `accessor<int, 1, access_mode::read_write, target::device>`
+- `local_accessor<int, 0>`
+- `unsampled_image_accessor<int, 1, access_mode::read>`
+- `sampled_image_accessor<int, 1>`
+- `stream`
+
+=== Host application tests
+
+These tests are run from a host application for all host types and kernel types.
+
+==== Equality via copy, symmetry
+
+This test is repeated for the copy constructor and for copy assignment. For class `T`:
+
+1. Create an instance `t0` of `T`.
+2. Create a copy of `t0`, `t1`.
+3. Check that `t0 == t1` and `t1 == t0`.
+4. Check that `std::hash<T>{}(t0) == std::hash<T>{}(t1)`.
+
+==== Non-equality via copy, symmetry
+
+This test is not executed for: `device_image`, `device` if only one device is available and `platform` if only one platform is available. For class `T`:
+
+1. Create an instance `t0` of `T`.
+2. Create an instance `t1` of `T`, which is distinct from `t0`.
+3. Check that `t0 != t1` and `t1 != t0`.
+4. Check that `std::hash<T>{}(t0) != std::hash<T>{}(t1)`.
+
+==== Reflexivity
+
+For class `T`:
+
+1. Create an instance `t0` of `T`.
+2. Check that `t0 == t0`.
+3. Check that `std::hash<T>{}(t0) == std::hash<T>{}(t0)`.
+
+==== Transitivity
+
+This test is repeated for the copy constructor and for copy assignment. For class `T`:
+
+1. Create an instance `t0` of `T`.
+2. Create a copy of `t0`, `t1`.
+3. Create a copy of `t1`, `t2`.
+4. Check that `t0 == t2` and `t2 == t0`.
+5. Check that `std::hash<T>{}(t0) == std::hash<T>{}(t2)`.
+
+=== Shared tests
+
+These tests are run both from a host application for all host types and kernel types, and from a kernel function for all kernel types. Some of the below tests require mutation of an instance `T`. This is not always possible. When it is not possible, do not execute those tests.
+
+==== Type traits
+
+For class `T`, check that the following statements evaluate to `true`.
+
+- `std::is_copy_constructible_v<T>`.
+- `std::is_copy_assignable_v<T>`
+- `std::is_destructible_v<T>`
+- `std::is_move_constructible_v<T>`
+- `std::is_move_assignable_v<T>`
+
+==== Copy
+
+This test is repeated for the copy constructor and for copy assignment. For class `T`:
+
+1. Create an instance `t0` of `T`.
+2. Store the properties of `t0` by calling its member functions.
+3. Create a copy of `t0`, `t1`.
+4. Check that the member functions of `t1` return the same values as stored in Step 2.
+
+==== Const-correctness
+Check that a `const` copy of an instance is not affected by the mutations to the instance that is copied.
+
+1. Create an instance `t0` of `T`.
+2. Make a `const` copy of `t0`, `t1`.
+3. Store the properties of `t1` by calling its member functions.
+4. Mutate `t0`.
+5. Check that the member functions of `t1` return the same values as stored in Step 2.
+
+==== Mutation propagation
+Check that mutations are applied both to the instance and to the copy of the instance.
+
+1. Create an instance `t0` of `T`.
+2. Make a copy of `t0`, `t1`.
+3. Mutate `t1`.
```

Comment:
What does "mutate" mean?  Step 4 below says "store properties", so I thought at first that "mutate" means "change the properties".  However, SYCL does not provide any way to change the properties of an object once it is created.  The common properties interface in [section 4.5.4.1](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_properties_interface) has only functions to query the properties (`has_property` and `get_property`).  There is no `set_property`.

Reply:
I replaced "properties" with "state" and clarified that the state of an object is the result of calling its member functions. The purpose of this is to check equality without relying on the presence of the equals operator.